CFE-2852: dnf_group package module

nickanderson opened the pull request:
Ticket: CFE-2852

nickanderson committed:
Added dnf_group package module for managing DNF package groups

This module enables CFEngine to manage DNF/YUM package groups (e.g.,
"Development Tools", "System Tools") on RHEL/Rocky/AlmaLinux systems.

Key features:
- Install, upgrade, and remove package groups
- List installed groups and check for updates
- Configure group installation types (mandatory/default/optional packages)
- Supports DNF setopt-style configuration options

Example usage:
  packages:
    "system-tools"
      policy => "present",
      package_module => dnf_group;

    "development"
      policy => "present",
      package_module => dnf_group,
      options => { "group_package_types=optional", "install_weak_deps=false" }
      version => "latest";  # Upgrade group packages

Ticket: CFE-2852

craigcomstock commented on modules/packages/vendored/dnf_group.mustache:
```diff
@@ -0,0 +1,556 @@
+                        sys.stdout.write(f"Architecture=all\n")
+                else:
+                    # TODO add a debug log for this case being triggered. or perhaps rip this out because we don't want to report that a group is installed just by the fact that one of it's packages are installed. that's not how dnf cli works. There seem to be default, mandatory and optional packages.
+                    # When i install system-tools it installs the default packages. if i install each of those packages indivudually system-tools is not shown to be installed.
```
```suggestion
                    # When i install system-tools it installs the default packages. if I install each of those packages individually system-tools is not shown to be installed.
```

craigcomstock commented on modules/packages/vendored/dnf_group.mustache:
```diff
@@ -0,0 +1,556 @@
+        # since the group was found and would be removed with proper setup
+        if "No such file or directory" in str(e):
+            logging.info(f"Transaction execution failed (likely cache issue): {str(e)}")
+            logging.info(f"Group would be removed with proper DNF setup")
```
What is proper dnf setup and improper? What would a person do to make it proper?

nickanderson commented on modules/packages/vendored/dnf_group.mustache:
```diff
@@ -0,0 +1,556 @@
+            else:
+                logging.error("No groups available.")
+            # Continue to next group instead of returning error to maintain idempotency
+            continue
```
they approved

nickanderson commented on modules/packages/vendored/dnf_group.mustache:
```diff
@@ -0,0 +1,556 @@
+            logging.error(f"Group '{group_name}' not found in available groups")
+            all_groups = list(base.comps.groups_iter())
+            if all_groups:
+                logging.error("Available groups:")
```
```
   error: package module: ErrorMessage=dnf package group nosystem-tools not found in available groups: backup-client, base-x, conflicts-appstream, container-management, debugging, desktop-debugging, dns-server, dotnet-core, ftp-server, gnome-apps, graphics, guest-agents, guest-desktop-agents, input-methods, internet-applications, internet-browser, java-platform, legacy-x, multimedia, office-suite, ostree-support, platform-kvm, platform-microsoft, platform-vmware, print-client, remote-desktop-clients, rpm-development-tools, texlive, virtualization-client, virtualization-hypervisor, virtualization-platform, virtualization-tools, web-server, additional-devel, anaconda-tools, base, conflicts-baseos, core, development, dial-up, file-server, fonts, gnome-desktop, graphical-admin-tools, hardware-monitoring, hardware-support, headless-management, infiniband, large-systems, legacy-unix, mail-server, mainframe-access, network-file-system-client, network-server, network-tools, networkmanager-submodules, performance, platform-devel, python-web, remote-system-management, scientific, security-tools, server-product, smart-card, smb-server, standard, system-tools, workstation-product
   error: Error installing package 'nosystem-tools'
```
I am not sure why `ErrorMessage=` is part of the error.

nickanderson commented:
I think that recursion failure is unrelated to this package module and instead related to that other file-separator change is that right @craigcomstock ?

larsewi commented:
> I think that recursion failure is unrelated to this package module and instead related to that other file-separator change is that right @craigcomstock ?

Try to rebase with upstream/master @nickanderson

larsewi commented on modules/packages/vendored/dnf_group.mustache:
```diff
@@ -0,0 +1,501 @@
+Dependencies: python3, python3-dnf
+"""
+
+from typing import Dict, List, Tuple, Optional, Any, Union
```
Remove unused imports
```suggestion
from typing import Dict, List, Tuple, Union
```

larsewi commented on modules/packages/vendored/dnf_group.mustache:
```diff
@@ -0,0 +1,501 @@
+"""
+
+from typing import Dict, List, Tuple, Optional, Any, Union
+import sys, logging, os, dnf
```
Multiple imports on one line

nickanderson commented:
> > I think that recursion failure is unrelated to this package module and instead related to that other file-separator change is that right @craigcomstock ?
> 
> Try to rebase with upstream/master @nickanderson

I have tried to rebase but it's still failing with the maximum recursion.

    info: Executing 'no timeout' ... '"/usr/bin/cf-agent" -f /var/lib/cfengine/inputs/update.cf --define bootstrap_mode '
  notice: Q: ".../cf-agent" -f /":    error: Maximum recursion level reached at '/var/lib/cfengine/inputs/cfe_internal/update/../../modules/packages/vendored'

nickanderson committed:
Added version-aware filter for vendored package paths

The filter() call in modules_presence needs to account for CFE-4623
behavior changes where findfiles() stopped suffixing directories with
trailing slashes in CFEngine 3.24.0+.

Prior to 3.24.0, findfiles() returned directory paths with trailing
slashes (e.g., "/path/vendored/"). Starting in 3.24.0, the trailing
slash was removed (e.g., "/path/vendored"). This broke the filter
pattern that was designed to exclude the vendored subdirectory.

The original fix (commit 3842469) unconditionally added $(const.dirsep)
to the filter pattern. However, this breaks on versions where findfiles()
still returns the trailing slash, causing maximum recursion errors during
policy updates.

This commit uses cf_version_between() and cf_version_at() to conditionally
add the directory separator only for versions affected by CFE-4623:
- 3.24.0 through 3.24.3
- 3.26.0
- 3.27.0

Versions outside this range use the original filter pattern without the
additional separator.

Related:
- CFE-4623: findfiles() should suffix directories with a slash
- Blog post: https://cfengine.com/blog/2025/change-in-behavior-findfiles/
- Original fix: commit 3842469153019413f727c6295b9e0a64247b07f6

Ticket: CFE-2852
Changelog: Fixed maximum recursion errors in modules_presence for CFEngine versions unaffected by CFE-4623
